File: Offer-Python/15.反转链表.py
## 题目描述
# 输入一个链表，反转链表后，输出新链表的表头。

# -*- coding:utf-8 -*-
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    # 返回ListNode
    def ReverseList(self, pHead):
        # write code here
        if not pHead or not pHead.next:
            return pHead
        last = None
        while pHead:
            temp = pHead.next
            pHead.next = last
            # last 和 pHead 向前移动
            last = pHead
            pHead = temp
        return last

File: Offer-Cpp/15.反转链表.cpp
//
// Created by 陈建虎 on 2019-11-04.
//
// 题目描述
// 输入一个链表，反转链表后，输出新链表的表头。

#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    struct ListNode *next;

    ListNode(int x) :
            val(x), next(NULL) {
    }
};

class Solution {
public:
    ListNode *ReverseList(ListNode *pHead) {
        if (pHead == NULL) {
            return NULL;
        }

        vector<ListNode *> nodes;
        while (pHead) {
            nodes.push_back(pHead);
            pHead = pHead->next;
        }

        for (int i = nodes.size() - 1; i > 0; --i) {
            nodes[i]->next = nodes[i - 1];
        }
        nodes[0]->next = NULL;

        return nodes.back();
    }
};